vowpalwabbit/core/src/reductions/epsilon_decay.cc

```diff
@@ -33,8 +35,129 @@ void epsilon_decay_score::update_bounds(float w, float r)
   _lower_bound = static_cast<float>(sd.first);
 }
 
+void epsilon_decay_score::reset_stats(double alpha, double tau)
+{
+  VW::scored_config::reset_stats(alpha, tau);
+  _lower_bound = 0.f;
+}
+
 float decayed_epsilon(uint64_t update_count) { return static_cast<float>(std::pow(update_count + 1, -1.f / 3.f)); }
 
+void epsilon_decay_data::update_weights(VW::LEARNER::multi_learner& base, VW::multi_ex& examples)
+{
+  auto num_models = static_cast<int64_t>(_scored_configs.size());
+  CB::cb_class logged{};
+  uint64_t labelled_action = 0;
+  const auto it =
+      std::find_if(examples.begin(), examples.end(), [](VW::example* item) { return !item->l.cb.costs.empty(); });
+  if (it != examples.end())
+  {
+    logged = (*it)->l.cb.costs[0];
+    labelled_action = std::distance(examples.begin(), it);
+  }
+
+  const float r = -logged.cost;
+  auto& ep_fts = examples[0]->_reduction_features.template get<VW::cb_explore_adf::greedy::reduction_features>();
+  // Process each model, then update the upper/lower bounds for each model
+  for (int64_t i = 0; i < num_models; ++i)
+  {
+    if (!_constant_epsilon)
+    { ep_fts.epsilon = VW::reductions::epsilon_decay::decayed_epsilon(_scored_configs[i][i].update_count); }
+    if (!base.learn_returns_prediction) { base.predict(examples, _weight_indices[i]); }
+    base.learn(examples, _weight_indices[i]);
+    for (const auto& a_s : examples[0]->pred.a_s)
+    {
+      if (a_s.action == labelled_action)
+      {
+        const float w = (logged.probability > 0) ? a_s.score / logged.probability : 0;
+        for (int64_t j = 0; j <= i; ++j) { _scored_configs[i][j].update_bounds(w, r); }
+        break;
+      }
+    }
+  }
+}
+
+// Promote model and all those lower with distance swap_dist
+void epsilon_decay_data::promote_model(int64_t model_ind, int64_t swap_dist)
+{
+  for (; model_ind >= 0; --model_ind)
+  {
+    for (int64_t score_ind = 0; score_ind < model_ind + 1; ++score_ind)
+    {
+      _scored_configs[model_ind + swap_dist][score_ind + swap_dist] = std::move(_scored_configs[model_ind][score_ind]);
+    }
+    std::swap(_weight_indices[model_ind + swap_dist], _weight_indices[model_ind]);
+  }
+}
+
+// Rebalance greater models to match lower shifted models
+void epsilon_decay_data::rebalance_greater_models(int64_t model_ind, int64_t swap_dist, int64_t num_models)
+{
+  int64_t greater_model = model_ind + swap_dist + 1;
+  for (int64_t curr_mod = greater_model; curr_mod < num_models; ++curr_mod)
+  {
+    for (int64_t score_ind = greater_model - swap_dist; score_ind > 0; --score_ind)
+    { _scored_configs[curr_mod][score_ind] = std::move(_scored_configs[curr_mod][score_ind - swap_dist]); }
```

Comment:
`std::move(_scored_configs[curr_mod][score_ind - swap_dist])`

The second index can become negative, which is undefined behavior

Reply:
Good catch, changed to `score_ind >= swap_dist`